### PR TITLE
[Timelock Partitioning] Part 48: Make use of templates for config whilst testing.

### PR DESCRIPTION
**Pull request by felixdesouza**
**Goals (and why)**:
There are quite a few configurations that we will end up testing. In the three node configuration, as it stands we have 3 separate configuration files for each node in the cluster and then for each minor modification/variation another set. This leads to configuration file rot where if we fix a bug/add useful logging/loggers, only a particular set of tests get it vs the other.

**Implementation Description (bullets)**:
* Use freemarker to do the templating.
* You pass in `TemplateVariables` to the cluster, it will then set up the right number of nodes, and generate the correct set of templates that does the right thing.

**Testing (What was existing testing like?  What have you done to improve it?)**:
* Existing tests should pass.

**Concerns (what feedback would you like?)**:
It's a bit "magic" but then again so is referencing random files and hoping they all work. It would be nice to serialise the config as is, but there's a lot of cruft which I don't think is serialisable and is also annoying to write, this feels like a good middle ground.

**Where should we start reviewing?**:
* `TemporaryConfigurationHolder`
* `TestableTimelockCluster`
* `TemplateVariables`
* `PaxosSuite`
* `paxosSingleServer.ftl`, `paxosMultiServer.ftl`

**Priority (whenever / two weeks / yesterday)**:
This week please.

### CODE REVIEW

File: timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/TemporaryConfigurationHolder.java [hunk 1]
```diff
@@ -16,59 +16,59 @@
...
+        Template template = TEMPLATE_CONFIG.getTemplate(templateName);
+        template.process(
+                variables.withDataDirectory(temporaryFolder.newFolder().getAbsolutePath()),
+                new FileWriter(temporaryConfigFile));
     }
 
     String getTemporaryConfigFileLocation() {
         return temporaryConfigFile.getPath();
     }
+
```
Comment (OStevan) [inline timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/TemporaryConfigurationHolder.java]: Is this intentional?

File: timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/TestableTimelockCluster.java [hunk 2]
```diff
@@ -56,35 +52,28 @@
...
+        this(name, configFileTemplate, ImmutableList.copyOf(variables));
     }
 
-    public TestableTimelockCluster(ClusterName clusterName, String baseUri, String... configFileTemplates) {
-        this.clusterName = clusterName.get();
-        this.configs = Arrays.stream(configFileTemplates)
-                .map(this::getConfigHolder)
+    public TestableTimelockCluster(String name, String configFileTemplate, Iterable<TemplateVariables> variables) {
+        this.name = name;
+        this.configs = Streams.stream(variables)
```
Comment (OStevan) [inline timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/TestableTimelockCluster.java]: Can we change this to `Set<TemporaryConfigurationHolder>`?

Comment (felixdesouza) [reply timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/TestableTimelockCluster.java]: any particular reason why?

Comment (felixdesouza) [reply timelock-server/src/testCommon/java/com/palantir/atlasdb/timelock/TestableTimelockCluster.java]: @OStevan ping on this

**Pull request merged**